x-pack/plugins/ml/public/application/jobs/new_job/pages/components/pick_fields_step/components/categorization_partition_field/categorization_per_partition.tsx

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { FC, useContext, useEffect, useState } from 'react';
+import { EuiFormRow } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { JobCreatorContext } from '../../../job_creator_context';
+import {
+  AdvancedJobCreator,
+  CategorizationJobCreator,
+  isCategorizationJobCreator,
+} from '../../../../../common/job_creator';
+import { newJobCapsService } from '../../../../../../../services/new_job_capabilities_service';
+import { Description } from './description';
+import { CategorizationPerPartitionSwitch } from './categorization_per_partition_switch';
+import { CategorizationPerPartitionFieldSelect } from './categorization_per_partition_input';
+import { CategorizationPerPartitionStopOnWarnSwitch } from './categorization_stop_on_warn_switch';
+
+export const CategorizationPerPartitionField: FC = () => {
+  const { jobCreator: jc, jobCreatorUpdate, jobCreatorUpdated } = useContext(JobCreatorContext);
+  const jobCreator = jc as AdvancedJobCreator | CategorizationJobCreator;
+  const [enablePerPartitionCategorization, setEnablePerPartitionCategorization] = useState(false);
+  const [categorizationPartitionFieldName, setCategorizationPartitionFieldName] = useState<
+    string | null
+  >(jobCreator.categorizationPerPartitionField);
+
+  const { catFields } = newJobCapsService;
+
+  const filteredCategories = catFields.filter((c) => c.id !== jobCreator.categorizationFieldName);
+
+  useEffect(() => {
+    jobCreator.categorizationPerPartitionField = categorizationPartitionFieldName;
+    jobCreatorUpdate();
+  }, [categorizationPartitionFieldName]);
+
+  useEffect(() => {
+    // set the first item in category as partition field by default
+    // because API requires partition_field to be defined in each detector with mlcategory
+    // if per-partition categorization is enabled
+    if (
+      jobCreator.perPartitionCategorization &&
+      jobCreator.categorizationPerPartitionField === null
+    ) {
+      jobCreator.categorizationPerPartitionField = filteredCategories[0].id;
```

Comment:
just to be safe, it would be worth adding a check for `filteredCategories.length` before access it.

Reply:
Updated in https://github.com/elastic/kibana/pull/74592/commits/a94ecaf246b7f84168eb296b5c2c1a75471353cc